ocaml/xapi/cert_distrib.ml

```diff
@@ -0,0 +1,262 @@
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "cert_distrib" end)
+
+module XenAPI = Client.Client
+
+type 'a remote_call =
+  API.ref_host -> (Rpc.call -> Rpc.response) -> API.ref_session -> 'a
+
+module Wire = struct
+  module Sexp = Sexplib.Sexp
+  open Sexplib0.Sexp_conv
+
+  type cert_blob = string [@@deriving sexp]
+
+  type cert = {uuid: string; blob: cert_blob} [@@deriving sexp]
+
+  type payload = Collect | Write of cert list | GenBundle [@@deriving sexp]
+
+  type result = CollectResult of cert_blob | WriteResult | GenBundleResult
+  [@@deriving sexp]
+
+  let string_of_payload x = x |> sexp_of_payload |> Sexp.to_string
+
+  let payload_of_string x =
+    try Some (x |> Sexp.of_string |> payload_of_sexp) with _ -> None
+
+  let dbg_of_payload = function
+    | Collect ->
+        "Collect"
+    | Write _ ->
+        "Write"
+    | GenBundle ->
+        "GenBundle"
+
+  let string_of_result x = x |> sexp_of_result |> Sexp.to_string
+
+  let result_of_string x =
+    try Some (x |> Sexp.of_string |> result_of_sexp) with _ -> None
+
+  let dbg_of_result = function
+    | CollectResult _ ->
+        "CollectResult"
+    | WriteResult ->
+        "WriteResult"
+    | GenBundleResult ->
+        "GenBundleResult"
+end
+
+let raise_internal ~details ~msg : 'a =
+  D.error "%s" details ;
+  raise Api_errors.(Server_error (internal_error, [msg]))
+
+(* eventually the remote calls should probably become API calls in the datamodel
+ * but they remain here for quick development *)
+module Worker : sig
+  val local_exec : __context:Context.t -> payload:string -> string
+
+  val remote_collect_cert : Wire.cert_blob remote_call
+
+  val remote_write_certs_fs : Wire.cert list -> unit remote_call
+
+  val remote_regen_bundle : unit remote_call
+end = struct
+  open Wire
+
+  let my_cert = "/etc/xensource/xapi-pool-tls.pem"
+
+  let pool_certs = "/etc/stunnel/certs-pool"
+
+  (* let bundle = "/etc/stunnel/xapi-pool-ca-bundle.pem" *)
+  let bundle = "/etc/stunnel/xapi-stunnel-ca-bundle.pem"
+
+  let pool_certs_bk = Printf.sprintf "%s.bk" pool_certs
+
+  let get_my_cert () =
+    let open Gencertlib.Pem in
+    match parse_file my_cert with
+    | Error reason ->
+        raise_internal
+          ~details:
+            (Printf.sprintf "get_my_cert: couldn't parse %s because %s" my_cert
+               reason)
+          ~msg:(Printf.sprintf "failed to parse %s" my_cert)
+    | Ok {host_cert} ->
+        host_cert
+
+  let mv a b =
```

Comment:
I think it would be better to implement this directly here. Likewise below for rm
```
(** [rmtree path] removes a file or directory recursively without following
      symbolic links. It may raise [Failure] *)
  let rmtree path =
    let ( // ) = Filename.concat in
    let rec rm path =
      let st = Unix.lstat path in
      match st.Unix.st_kind with
      | Unix.S_DIR ->
          Sys.readdir path |> Array.iter (fun file -> rm (path // file)) ;
          Unix.rmdir path
      | _ ->
          Unix.unlink path
```

Reply:
See https://github.com/xapi-project/xenopsd/blob/master/lib/xenops_utils.ml#L161-L203 - I would expect similar code in xapi but don't know from the top of my head.

Reply:
I've added rmrf and mv to helpers.ml (implemented in ocaml) - I couldn't see these functions in xapi already, which is a bit surprising